### ISSUE TITLE
Mipmaps Limit import setting is not respected

### ISSUE THREAD

**Issue by allenwp** [godotengine/godot#63934]
### Godot version

v4.0.alpha13.official [59dcddc45]

### System information

Windows 10, Vulkan Mobile, NVIDA 980 Ti

### Issue description

The Mipmaps Limit setting in the Texture2D import settings is not respected.

![mipmaps-limit-bug](https://user-images.githubusercontent.com/17506573/182982305-61d01c76-5282-4b77-bbe5-eed93453cf69.png)

(In the above screenshot, note that the Limit is set to 5, but there are 13 mipmaps for this Texture2D resource.)

### Steps to reproduce

1. Add a new Texture2D resource to the FileSystem.
2. Select the new texture resource and view the Import panel
3. Turn on Generate under Mipmaps
4. Click "Reimport" button
5. Set the Limit to something other than -1
6. Click the "Reimport" button again
7. Select the Texture2D resource in the FileSystem to view it in the Inspector panel
8. Note that the inspector shows that there are more mipmaps in the Texture2D resource than in the Limit setting

### Minimal reproduction project

[MipmapsLimitBug.zip](https://github.com/godotengine/godot/files/9264492/MipmapsLimitBug.zip)

**Comment by Chaosus** [comment]
Does it reproducible in 3.5?

**Comment by allenwp** [comment]
> Does it reproducible in 3.5?

The "Limit" setting does not exist in v3.5.stable.official [991bb6ac7], so this bug is specific to 4.0.

**Comment by allenwp** [comment]
From looking at the source code, it appears that this Limit setting is not yet used by the `Image` class when mipmaps are generated. In `ResourceImporterTexture::_save_ctex`, the setting is saved to the CTEX file: `f->store_32(p_limit_mipmap);`, but then is never passed on to the `Image` for it to be used by its `Image::generate_mipmaps` function.

Further to this, `Image::_get_dst_image_size` seems to just count to the max possible mipmaps without using the Limit setting.

**Comment by Calinou** [comment]
I can still reproduce this on 4.0.2.

**Comment by vfig** [comment]
this bit me today, in 4.1.3.

i was rendering grass blades, and having issues with a whitish fringe appearing when the blades were at a steep angle to the camera, or the blades turning entirely to white further from the camera.
![grassbug](https://github.com/godotengine/godot/assets/81129/f6c22524-a4f2-47a2-b685-8b1935628941)
![grassbug2](https://github.com/godotengine/godot/assets/81129/e679e961-9c10-4a6e-bccd-c406499dffb1)

the grass is being rendered with the texture below. using renderdoc, i concluded the issue was due to the 7th (1x1) mip being used, as by then the texture is reduced to a single whitish-green pixel.
![grass_mip0](https://github.com/godotengine/godot/assets/81129/94393695-0ce7-4d20-abb6-15090ba76c0e)
![grass_mip7_zoomed](https://github.com/godotengine/godot/assets/81129/827d08b9-d1ea-4d4f-a3ac-f2e9f0463963)

i tried to use the Mipmaps Limit setting on the .tga's import settings and reimporting, and found it was having no effect at all. this misled me into thinking the fringing problem was something other than mips, but after some time i eventually changed the shader to manually specify the lod to sample, to confirm that yes, it was actually this mip being the problem, and that for some reason the Mipmaps Limit setting was not working.

for a workaround, i suppose i can either generate a .dds elsewhere without this last mip, or query and clamp the mip levels in the fragment shader.

**Comment by NuclearPhoenixx** [comment]
I ran into a problem today where textured terrain in the distance looked quite weird and distorted due to mipmaps (nearest mipmap filter on the texture). My best guess is this is caused by the last one or two mipmaps that get generated on my pixel art texture leading to a miscoloring of terrain.

Anyways, I don't want to got into too much detail here, it would be extremely helpful to limit the number of mipmaps so I support this issue. Is there any other way to maybe don't generate the last mipmap or so?

**Comment by Calinou** [comment]
>  Is there any other way to maybe don't generate the last mipmap or so?

Yes, using a DDS texture that comes with custom mipmaps already present in the file.

**Comment by Calinou** [comment]
I've started looking at fixing this issue, but I couldn't get it to work yet as uninitialized data is present in the smaller mipmaps once you turn on the limit: https://github.com/Calinou/godot/tree/image-fix-mipmaps-limit-import-option

**Testing project:** [test_rgss.zip](https://github.com/user-attachments/files/16074327/test_rgss.zip)

![image](https://github.com/godotengine/godot/assets/180032/aa7e3533-3419-4635-89b4-896c7adfd1af)

**Comment by NuclearPhoenixx** [comment]
> * Related to [Add option to "Preserve alpha test coverage" on mip map generation for texture import godot-proposals#9152](https://github.com/godotengine/godot-proposals/issues/9152).
> 
> 
> > Is there any other way to maybe don't generate the last mipmap or so?
> 
> Yes, using a DDS texture that comes with custom mipmaps already present in the file. Many programs like GIMP can write DDS textures with custom mipmaps (one mipmap in each layer).

Revisiting this issue... It seems like Godot only likes DDS files with a fixed number of 8 mipmaps.

Using texconv to generate only 3 mipmaps, for example, results in Godot complaining:

```
ERROR: Expected Image data size of 256x256x1 (BPTC_RGBA with 8 mipmaps) = 87408 bytes, got 86016 bytes instead.
ERROR: Invalid image: image is empty
ERROR: Failed loading resource: res://textures/test.dds. Make sure resources have been imported by opening the project in the editor at least once.
```

**Comment by Calinou** [comment]
@NuclearPhoenixx This may be addressed by https://github.com/godotengine/godot/pull/108720, can you [test it locally](https://docs.godotengine.org/en/stable/contributing/workflow/testing_pull_requests.html)?

**Comment by NuclearPhoenixx** [comment]
Tested and looks like to be working, nothing unexpected happening. Sweet!

**Comment by SK83RJOSH** [comment]
Just a heads-up that any solution for this issue should handle cases like `load_dds_from_buffer`. This hit me this week, and my assets come from external sources that I do not control (it's a mod tool for an older game). The resource loaders ironically handle it perfectly, but get hit by this oversight when initializing the image.
